src/algebraic_geometry/projective_spectrum/structure_sheaf.lean

```diff
@@ -0,0 +1,218 @@
+`f : Π x : U, homogeneous_localization x` such that `f` is locally expressible as ratio of two
+elements of the *same grading*, i.e. `∀ y ∈ U, ∃ (V ⊆ U) (i : ℕ) (a b ∈ 𝒜 i), ∀ z ∈ V, f z = a / b`.
+
+* `algebraic_geometry.projective_spectrum.structure_sheaf.is_locally_fraction`: the predicate that
+  a dependent function is locally expressible as ration of two elements of the same grading.
+* `algebraic_geometry.projective_spectrum.structure_sheaf.sections_subring`: the dependent functions
+  satisfying the above local property forms a subring of all dependent functions
+  `Π x : U, homogeneous_localization x`.
+* `algebraic_geometry.Proj.structure_sheaf`: the sheaf with `U ↦ sections_subring U` and natural
+  restriction map.
+
+Then we establish that `Proj 𝒜` is a `LocallyRingedSpace`:
+* `algebraic_geometry.homogeneous_localization.is_local`: for any `x : projective_spectrum 𝒜`,
+  `homogeneous_localization x` is a local ring.
+* `algebraic_geometry.Proj.stalk_iso'`: for any `x : projective_spectrum 𝒜`, the stalk of
+  `Proj.structure_sheaf` at `x` is isomorphic to `homogeneous_localization x`.
+* `algebraic_geometry.Proj.to_LocallyRingedSpace`: `Proj` as a locally ringed space.
+
+## References
+
+* [Robin Hartshorne, *Algebraic Geometry*][Har77]
+
+
+-/
+
+noncomputable theory
+
+namespace algebraic_geometry
+
+open_locale direct_sum big_operators pointwise
+open direct_sum set_like
+
+variables {R A: Type*}
+variables [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ℕ → submodule R A) [graded_algebra 𝒜]
+
+local notation `at ` x := localization.at_prime x.as_homogeneous_ideal.to_ideal
+
+open Top topological_space category_theory opposite
+
+/--
+The underlying topology of `Proj` is the projective spectrum of graded ring `A`.
+-/
+def projective_spectrum.Top : Top := Top.of (projective_spectrum 𝒜)
+
+namespace projective_spectrum.structure_sheaf
+
+namespace homogeneous_localization
+
+open set_like.graded_monoid submodule
+
+variables {𝒜} {x : projective_spectrum.Top 𝒜}
+
+/--
+If `x` is a point in `Proj 𝒜`, then `y ∈ Aₓ` is said to satisfy `num_denom_same_deg` if and only if
+`y = a / b` where `a` and `b` are both in `𝒜 i` for some `i`.
+-/
+@[nolint has_inhabited_instance]
+structure num_denom_same_deg (y : at x) :=
+(num denom : A)
+(denom_not_mem : denom ∉ x.as_homogeneous_ideal)
+(deg : ℕ)
+(num_mem : num ∈ 𝒜 deg)
+(denom_mem : denom ∈ 𝒜 deg)
+(eq : (localization.mk num ⟨denom, denom_not_mem⟩ : at x) = y)
+
+attribute [simp] num_denom_same_deg.eq
+
+variable (x)
+/--
+Auxiliary definition of `homogeneous_localization`: its underlying set.
+-/
+def carrier : set (at x) :=
+{y | nonempty (num_denom_same_deg y)}
+
+variable {x}
+lemma one_mem' : (1 : at x) ∈ carrier x := nonempty.intro
+{ num := 1,
+  denom := 1,
+  denom_not_mem := (ideal.ne_top_iff_one _).mp x.is_prime.ne_top,
+  deg := 0,
+  num_mem := one_mem,
+  denom_mem := one_mem,
+  eq := by simp }
```

Comment:
This probably makes sense as a `has_one (num_denom_same_deg y)` instance, similarly below